### ISSUE TITLE
False positive with PEP563 subscripted generics

### ISSUE THREAD

**Issue by patrick-kidger** [beartype/beartype#218]
```python
from __future__ import annotations

from beartype import beartype
from typing import Generic, TypeVar

T = TypeVar("T")

class Foo(Generic[T]):
    @beartype
    def bar(self) -> Foo[T]:
        return self

Foo().bar()
# beartype.roar.BeartypePep563Exception: @beartyped __main__.Foo.bar() return PEP 563-postponed type hint 'Foo[T]' syntactically invalid (i.e., "name 'Foo' is not defined") under:
```
:(

**Comment by patrick-kidger** [comment]
The same also occurs with PEP563-annotations + Python 3.10 union syntax: `-> None | Foo`.

**Comment by leycec** [comment]
**Ah-ha!** I can't believe my lying eyes here, but... @beartype already supports this. The error message could arguably be improved to suggest the solution, which is to decorate your *class* `Foo` rather than your *method* `bar()` by `@beartype`. At the early decoration time that `@beartype` runs when applied to `bar()`, **`Foo` has yet to be declared.**

This is what the explanatory substring `"name 'Foo' is not defined"` is trying (but clearly failing) to convey. If you decorate `Foo` instead, this actually works:

```python
from __future__ import annotations

from beartype import beartype
from typing import Generic, TypeVar

T = TypeVar("T")

@beartype
class Foo(Generic[T]):
    def bar(self) -> Foo[T]:
        return self

Foo().bar()
```

:muscle: :bear: :muscle: 

Please do reopen this if there's something here we've missed. And thanks so much for the recent glut of ugly issues! It feels bad but compels us to do something, which is good.

```pycon
>>> bad + good = good
True
```

**Comment by patrick-kidger** [comment]
Oh, interesting!

But -- why does this affect anything? Using PEP 563 (evil as it is), then all annotations are automatically deferred. Why is this any different to
```python
from __future__ import annotations

from beartype import beartype

@beartype
def f(x: Foo):
    pass

class Foo:
    pass

```
?

So I'm guessing that it is possible to support the decoration style from my original issue? (E.g. if you want to decorate some methods but not others.)

**Comment by leycec** [comment]
**I love these questions.** Google X: direct more grant funding towards this infamous intellect, because he is worth his weight in gold-backed cryptocurrency.

You have this uncanny knack for getting to the smelly heart of things. Let's see... Taking things from the top:

> But -- why does this affect anything? Using PEP 563 (evil as it is), then all annotations are automatically deferred.

...right. There's only one problem with that convenient theory: PEP 563 was never intended to be supported at runtime. It was only ever a syntactic convenience for codebases subject to static type-checking. Unsurprisingly, breaking runtime in favour of static analysis time was a bad idea. But the bad idea persists to this day. Let's whack it once more!

So, @beartype basically has two orthogonal modes of operation here:

* **PEP 484 mode,** AKA what @beartype *usually* does. In this mode, @beartype currently treats stringified type hints as **forward references** to classes that (typically) have yet to be declared. This is the original interpretation of stringified type hints as standardized by PEP 484. But your recent flood of incisive issues demonstrates that that's not enough. Everybody really wants stringified type hints to be **arbitrarily complex hints** intended to be dynamically evaluated by the `eval()` builtin against the local and global attributes defined by the local and global scopes encapsulating those type hints. Everybody: you are not wrong. This is the goal we strive towards, too.
* **PEP 563 mode,** enabled with the `from __future__ import annotations` pragma. In this mode, @beartype treats *all* type hints as strings intended to be dynamically evaluated by the `eval()` builtin against the local and global attributes defined by the local and global scopes encapsulating those type hints. Order of operations is significant here. In fact, @beartype strives to achieve (and mostly succeeds in achieving) perfect resolution of PEP 563-postponed type hints. After evaluation, *most* of the resulting type hints will no longer be strings; they'll be actual usable type hints. This is the exact same approach outlined by [section "Resolving Type Hints at Runtime" of PEP 563](https://peps.python.org/pep-0563/#resolving-type-hints-at-runtime) and naively implemented by the standard `typing.get_type_hints()` function:

> To resolve an annotation at runtime from its string form to the result of the enclosed expression, **user code needs to evaluate the string.**
>
> For code that uses type hints, the `typing.get_type_hints(obj, globalns=None, localns=None)` function correctly evaluates expressions back from its string form. Note that all valid code currently using `__annotations__` should already be doing that since a type annotation can be expressed as a string literal.
>
> For code which uses annotations for other purposes, a regular `eval(ann, globals, locals)` call is enough to resolve the annotation.

Superficially, they appear similar. In actuality, they differ so extremely that this might be worth describing in several FAQ entries altogether comprising a mind-numbing dissertation on this subject as well as the psychosocial nature of man vis-a-vis type hints.  

Basically, `from __future__ import annotations` doesn't do what you want, because PEP 563 wasn't standardized to do what you want, because PEP 563 is dumb. You want PEP 563 to implicitly defer the evaluation of all type hints to the last possible time at runtime (e.g., to the time at which a callable annotated by those type hints is **called** rather than merely **decorated** by a runtime type-checker), right?

But PEP 563 doesn't care about runtime. PEP 563 only cares about static type-checkers, because those were the only type-checkers thought to matter when PEP 563 was drafted. So PEP 563 just instructs runtime type-checkers (all necessarily implemented as decorators and thus run at decoration time) to `eval()` everything immediately.

Even if @beartype *wanted* to defer the evaluation of PEP 563-postponed type hints to call time rather than decoration time, we couldn't. @beartype dynamically generates highly performant type-checking code. This makes @beartype *extremely* slow at decoration time but comparably fast at call time. If @beartype didn't `eval()` PEP 563-postponed type hints at decoration time, @beartype wouldn't be able to generate *any* type-checking code whatsoever under PEP 563. Generating something is better than generating nothing – and nobody should be really using PEP 563 anyway in 2023, because doing so substantially impacts runtime performance.

So, @beartype just does what PEP 563 tells us. Which is dumb, because this means that static type-checkers will accept code rejected by runtime type-checkers when PEP 563 is active. From the runtime perspective, PEP 563 had might as well not exist at all. It's just an expensive noop at runtime.

## tl;dr

PEP 563 is an expensive noop at runtime. I could have just said that. Instead, I wrote a bad Master's Thesis up there and now I have nothing to show for it. :face_exhaling: 

But then...

> Why is this any different to

```python
from __future__ import annotations

from beartype import beartype

@beartype
def f(x: Foo):
    pass

class Foo:
    pass
```

> ?

**Oh, boy.** You just discovered something you shouldn't have. If this was a video game, you would have just gotten +999999XP and the "Unquenchable Thirst" Achievement. Since this is reality, you only get @leycec's solemn fist bump. :fist_right: :fist_left: 

If you replace the signature `def f(x: Foo):` with `def f(x: Foo[int]):`, @beartype will raise the expected exception that `Foo` has yet to be defined. But... why doesn't @beartype do that in the above edge case, too?

*...heh.* The answer is that PEP 563 is dumb. PEP 563 never provided any runtime support in *any* way whatsoever. In fact, **PEP 563 never even provided an API for detecting PEP 563.** To see whether or not a module has PEP 563 enabled, @beartype quietly invades the privacy of your module and looks for an import attribute named `annotations` whose value is the `__future__.annotations` singleton.

So what's the problem here? @beartype has no way of knowing whether any competing runtime type-checker has already resolved PEP 563-postponed type hints for this module. After all, it's not infeasible that a user could combine @beartype *and* Pydantic (which is quite popular in webdev circles). If Pydantic runs first and supports PEP 563, then Pydantic will almost certainly resolve *all* of the PEP 563-postponed type hints that it can – because it has to to do its job. Right? And @beartype has no means of distinguishing these two cases. Repeat after me: "PEP 563 is dumb."

So what should @beartype do when it sees a type hint that is a Python identifier under PEP 563? That type hint *could* be a PEP 563-postponed type hint that has yet to be resolved. Alternately, that type hint *could* be a forward reference to an undefined class that has already been resolved by a competing runtime type-checker. Syntactically, they're identical strings in both cases. Semantically, they're treated completely different. In the former case, the string requires passing to `eval()`; in the latter case, it doesn't.

For safety, @beartype prefers the latter approach. After all, there's no harm and considerable gain to just preserving forward references as is. Right? But then this *super*-weird edge case arises that nobody's supposed to know about.

Of course, it's actually a bit more nuanced than even that deep dive above. I'll just copy-and-paste myself from internal commentary in our PEP 563 resolution algorithm:

```python
            # If this hint is the unqualified name of one or more parent
            # callables or classes of this callable, then this hint is a
            # relative forward reference to a parent callable or class of this
            # callable that is currently being defined but has yet to be
            # defined in full. By deduction, we can infer this hint *MUST* have
            # been a locally or globally scoped attribute of this callable
            # before being postponed by PEP 563 into a relative forward
            # reference to that attribute: e.g.,
            #     # If this loop is iterating over a postponed type hint
            #     # annotating this post-PEP 563 method signature...
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> 'MuhClass': ...
            #
            #     # ...then the original type hints prior to being postponed
            #     # *MUST* have annotated this pre-PEP 563 method signature.
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> MuhClass: ...
            #
            # In this case, we absolutely *MUST* avoid attempting to resolve
            # this forward reference. Why? Disambiguity. Although the
            # "MuhClass" class has yet to be defined at the time @beartype
            # decorates the muh_method() method, an attribute of the same name
            # may already have been defined at that time: e.g.,
            #     # While bad form, PEP 563 postpones this valid logic...
            #     MuhClass = "Just kidding! Had you going there, didn't I?"
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> MuhClass: ...
            #
            #     # ...into this relative forward reference.
            #     MuhClass = "Just kidding! Had you going there, didn't I?"
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> 'MuhClass': ...
            #
            # Naively resolving this forward reference would erroneously
            # replace this hint with the previously declared attribute rather
            # than the class currently being declared: e.g.,
            #     # Naive PEP 563 resolution would replace the above by this!
            #     MuhClass = "Just kidding! Had you going there, didn't I?"
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> (
            #             "Just kidding! Had you going there, didn't I?"): ...
            #
            # This isn't simply an edge-case disambiguity, however. This exact
            # situation commonly arises whenever reloading modules containing
            # @beartype-decorated callables annotated with self-references
            # (e.g., by passing those modules to the standard
            # importlib.reload() function). Why? Because module reloading is
            # ill-defined and mostly broken under Python. Since the
            # importlib.reload() function fails to delete any of the attributes
            # of the module to be reloaded before reloading that module, the
            # parent callable or class referred to by this hint will be briefly
            # defined for the duration of @beartype's decoration of this
            # callable as the prior version of that parent callable or class!
            #
            # Resolving this hint would thus superficially succeed, while
            # actually erroneously replacing this hint with the prior rather
            # than current version of that parent callable or class. @beartype
            # would then wrap the decorated callable with a wrapper expecting
            # the prior rather than current version of that parent callable or
            # class. All subsequent calls to that wrapper would then fail.
            # Since this actually happened, we ensure it never does again.
            #
            # Lastly, note that this edge case *ONLY* supports top-level
            # relative forward references (i.e., syntactically valid Python
            # identifier names subscripting *NO* parent type hints). Child
            # relative forward references will continue to raise exceptions. As
            # resolving PEP 563-postponed type hints effectively reduces to a
            # single "all or nothing" call of the low-level eval() builtin
            # accepting *NO* meaningful configuration, there exists *NO* means
            # of only partially resolving parent type hints while preserving
            # relative forward references subscripting those hints. The
            # solution in those cases is for end users to either:
            #
            # * Decorate classes rather than methods: e.g.,
            #     # Users should replace this method decoration, which will
            #     # fail at runtime...
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> list[MuhClass]: ...
            #
            #     # ...with this class decoration, which will work.
            #     @beartype
            #     class MuhClass:
            #         def muh_method(self) -> list[MuhClass]: ...
            # * Replace implicit with explicit forward references: e.g.,
            #     # Users should replace this implicit forward reference, which
            #     # will fail at runtime...
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> list[MuhClass]: ...
            #
            #     # ...with this explicit forward reference, which will work.
            #     class MuhClass:
            #         @beartype
            #         def muh_method(self) -> list['MuhClass']: ...
            #
            # Indeed, the *ONLY* reasons we support this common edge case are:
            # * This edge case is indeed common.
            # * This edge case is both trivial and efficient to support.
            #
            # tl;dr: Preserve this hint for disambiguity and skip to the next.
```

And then, much later:

```python
            # Attempt to resolve this hint against both the global and local
            # scopes for the decorated callable.
            try:
                func_hints_resolved[pith_name] = eval(
                    pith_hint, func_globals, func_locals)
            # If that resolution also fails...
            except Exception as exception:
                # If...
                if (
                    # That resolution fails with a "NameError" *AND*...
                    isinstance(exception, NameError) and
                    # This hint is a valid Python identifier...
                    is_identifier(pith_hint)
                ):
                    # This hint is *PROBABLY* a forward reference hinted as a
                    # string. In this case, defer validation of this string as a
                    # valid forward reference to a class (which presumably has
                    # yet to be declared) until call time of the decorated
                    # callable by preserving this string as is.
                    #
                    # PEP 563 prevents runtime type checkers from distinguishing
                    # between forward references hinted as strings and
                    # non-forward references postponed under PEP
                    # 563 as strings. Ideally, PEP 563 would postpone the former
                    # as machine-readable string representations (e.g.,
                    # converting "muh.class.name" to "'muh.class.name'"). It
                    # doesn't. Instead, it simply preserves forward references
                    # hinted as strings! Who approved this appalling
                    # abomination that breaks CPython itself?
                    # print(f'Deferring postponed forward reference hint {repr(pith_hint)}...')
                    continue
                # Else, this hint is *PROBABLY NOT* a forward reference hinted
                # as a string.
```

Note the trailing admonition: "Who approved this appalling abomination that breaks CPython itself?"

Whenever you enable PEP 563, the answer is always: "Because PEP 563 is busted."

> Might it still be possible to support the decoration style from my original issue? (E.g. if you want to decorate some methods but not others.)

Yes, yes, YES!!! `</ahem>`

These are *such* good questions, @patrick-kidger. You are on a typing roll. Show the brilliant researcher what he's won tonight, Emoji Cat! :cat2: 

It's... already working support for this exceptional question! @beartype supports a blacklisting approach here, enabling you to selectively exclude specific methods from type-checking. Because you can never have enough bad choices on the table, @beartype gives you two choices here. Just decorate methods to be excluded by either:

* The [standard `@typing.no_type_check` decorator](https://docs.python.org/3/library/typing.html#typing.no_type_check).
* The [@beartype-specific `BeartypeStrategy.O0` strategy](https://beartype.readthedocs.io/en/latest/api_decor/#beartype.BeartypeStrategy.O0).

In theory, @beartype should transparently support both. `@typing.no_type_check` is standard, but less widely applicable; `BeartypeStrategy.O0` is non-standard, but can be used wherever @beartype accepts a `conf: BeartypeConf` parameter (e.g., the statement-level `beartype.door.is_bearable()` runtime type-checker). Here's a minimal-length example of the latter, 'cause I've already depleted enough of everyone's precious life force today:

```python
# Import the requisite machinery.
from beartype import beartype, BeartypeConf, BeartypeStrategy

# Dynamically create a new @nobeartype decorator disabling type-checking.
nobeartype = beartype(conf=BeartypeConf(strategy=BeartypeStrategy.O0))

# Automatically decorate all methods of this class...
@beartype
class TypeCheckedClass(object):
    # Including this method, which raises a type-checking violation
    # due to returning a non-"None" value.
    def type_checked_method(self) -> None:
        return 'This string is not "None". Apparently, that is a problem.'

    # Excluding this method, which raises *NO* type-checking
    # violation despite returning a non-"None" value.
    @nobeartype
    def non_type_checked_method(self) -> None:
        return 'This string is not "None". Thankfully, no one cares.'
```

I think we've even tested this somewhere. Maybe. :joy_cat:

**Comment by patrick-kidger** [comment]
> I love these questions. Google X: direct more grant funding towards this infamous intellect, because he is worth his weight in gold-backed cryptocurrency.

Haha, thank you -- that gave me a good laugh.

On to the technical stuff:

1. When hints are evaluated;

2. PEP 563 vs strings.

It seems to me that basically all of the difficulty stems from a desire to early-evaluate hints. This approach seems pretty fraught with issues, e.g. the `importlib.reload` stuff you mention, the issue I've raised here (and some of the others I've recently mentioned), etc.

What about just always deferring resolving type hints to runtime (which I think is always correct under all circumstances; this follows the usual late-binding behaviour of Python closures), and then just caching the resolved hints after the first invocation?

If you're really slick you might even be able to do this by modifying the function binding in the stack frame above, so that subsequent invocations automatically get the resolved version, without even the overhead of an `if resolve` check.

3. Competing resolvers

> So what's the problem here? @beartype has no way of knowing whether any competing runtime type-checker has already resolved PEP 563-postponed type hints for this module. 

I'm not sure I understand -- what do you mean here by "resolve" the hint, i.e. why should beartype care what pydantic does? Either you get a string that needs to be `eval`'d (subject to `importlib.reload` chicanery), or the actual hint itself. Hopefully neither of you are writing back to the `__annotations__` dictionary, which would be a bad-citizen sort of thing to do. (And even if one of you is, you still only get a string or the actualy resolved hint, which is the same situation as before, so there still isn't a problem.)
